### DISCUSSION TITLE
Char LCD 20x4 via I2C does not work

### DISCUSSION THREAD

**Discussion by HJLK** [rern/rAudio#1285] [Bug Reports]
My character LCD (HD44780) does not work, it is not even initialized at boot. 

INSTALL: Character LCD 20 x 4 via PCF8574-chip (address 0x27) coupled on RPi3 I2C pins (board pin 3 = SDA, pin 5 = SCL)

This is the content of the conf-file srv/http/data/system/lcdchar.conf

**[var]
cols=20
charmap=A00
inf=i2c
address=39
chip=PCF8574
backlight=False**

The lcdchar-file is empty (0kB).

Am I missing something? Did I make an error?

**Comment by HJLK** [comment]
NB: I deactivated the POWER-button (I2C uses pin5 as clock).

**Comment by HJLK** [comment]
i2cdetect show the char LCD is known on address 0x27:


rAudio:~  #  **i2cdetect -y 1**
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:                         -- -- -- -- -- -- -- --
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
20: -- -- -- -- -- -- -- 27 -- -- -- -- -- -- -- --
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
40: -- -- -- -- -- -- -- -- -- -- -- -- -- UU -- --
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
70: -- -- -- -- -- -- -- --
rAudio:~  #

**Reply by HJLK** [reply]
Sorry! I overlooked the 5V-3.3V - problem in the "link" in the screen... (RTFM!). I will follow that, and report the results.

**Reply by HJLK** [reply]
I followed the instruction. Feed 5V to Vdd of LCD, after cutting the connection to the backpack. Feed 3.3V tot the backpack. There was no difference: still nothing to see, i2cdetect showed 27 as address. So again, am I missing something?

**Reply by HJLK** [reply]
So I tried to address the LCD thru python:

>>> from RPLCD.i2c import CharLCD
>>> lcd = CharLCD(i2c_expander='PCF8574', address=0x27, port=1,
...               cols=20, rows=4, dotsize=8,
...               charmap='A02',
...               auto_linebreaks=True,
...               backlight_enabled=True)
>>> lcd.write_string('Hello world')
>>>

And lo and behold: it is working! The LCD shows "Hello world"... So I2C is working and the hardware is OK. 
So again: what am I missing?

**Reply by HJLK** [reply]
More information: I opened a python session, and executed lcdcharconfig.py manually. The lcd did initialize and subsequent write_strings() showed on the lcd. My conclusion: the preps are done and ok, but **the system does NOT call lcdcharconfig.py**. So me thinks it's a bug.
To be very sure I rebuilt a new microSD-card from the last image and updated it to 20230310, and this made no difference.

**Reply by HJLK** [reply]
I dove (and went nearly under) into the charLCD sotfware... and this is difficult, because there is hardly logging.
1) in /srv/http/bash I executed first "lcdcharcinfig.py" and then "lcdchar.py logo". The LCD initialized, and showed rAudio but NOT the rAudio logo (it showed in stead the default chars in the LCD). **Conclusion**: "lcdcharinit.py" is NEVER executed. 
Systemd runs once per boot "startup.sh", in which I see:

_if [[ -e $dirsystem/lcdchar ]]; then
	lcdcharinit.py
	lcdchar.py logo
fi_

So this part of startup.sh does not work. Why? Logging of systemd shows:

_Apr 16 11:16:12 rAudio startup.sh[730]: /srv/http/bash/status.sh: line 22: /srv/http/data/shm/player: No such file or directory
Apr 16 11:16:14 rAudio startup.sh[876]: /srv/http/bash/startup.sh: line 116: lcdcharinit.py: command not found
Apr 16 11:16:14 rAudio startup.sh[877]: /srv/http/bash/startup.sh: line 117: lcdchar.py: command not found
_

 So there seems to be problem...

2) In lcdchar.py there is another problem:

_sys.path.append( '/srv/http/data/shm' )
from statuslcd import *_

Here I see an error: **mpd** does not exist. It is the last line of status: player=mpd. There is no definition of the variable mpd in lcdchar.py.

So far my investigations. I hope this is enough to resolve my problems.

**Comment by HJLK** [comment]
My CharLCD 20x4 is working now, with some restrictions. However, I had to modify 3 files, all in /srv/http/bash: startup.sh, status-push.sh and lcdchar.py. 

1) In **startup.py** I added $dirbash/ to the .py's in lines 116, 117. The systemd log showed that these .py's "did not exist". This way they were found, and the charLCD was initialized, and I saw the initial logo (nice work Rern!). 

if [[ -e $dirsystem/lcdchar ]]; then
	$dirbash/lcdcharinit.py     # HJLK $dirbash/ ontbrak voor de .py
	$dirbash/lcdchar.py logo    # HJLK $dirbash/ ontbrak voor de .py
fi

2) In **status-push.sh** the file /dev/shm/status is "copied" to the file  **statuslcd.py**, which is then imported in lcdchar.py. The import crashed because of missing quotes around values **mpd** (player=), **play**, **stop**, and **pause** (state=). For bash that makes no difference, but python hates it: "mpd is not defined". So to leave **status** to itself, I added 4 lines to "quote" these values in statuslcd.py,  in **status-push.sh** (lines 56-59). Also added $dirbash/ to lcdchar.py. OK, it can be more efficient I think, but "keep it simple and stupid" is my motto (one of them, anyway).

_if [[ -e $dirsystem/lcdchar ]]; then
	sed -E 's/(true|false)$/\u\1/' $dirshm/status > $dirshm/statuslcd.py
	sed -i 's/state=stop/state="stop"/'       $dirshm/statuslcd.py
	sed -i 's/state=pause/state="pause"/'     $dirshm/statuslcd.py
	sed -i 's/state=play/state="play"/'       $dirshm/statuslcd.py
	sed -i 's/=mpd/="mpd"/'                   $dirshm/statuslcd.py
	$dirbash/lcdchar.py &> /dev/null &_

3) In lcdchar.py at last there is a dictionary **icon** defined. That clashes with the later import of statuslcd.py, where sometimes a line **icon="picture"** is present. Which leads to an error "stringindex must be integer" in **icon[ state ]**, later. (That was a hard one, it took me hours to find that!). This can be avoided by renaming the dictionary to **status_icon**. All worked then, except the "4th line (with the time ticking). I commented this part, because I think this way it is not possible: lcdchar.py runs "forever" in the background the first time it is started. Subsequent runs will result in more occurrences that access the lcd, which will clobber the lcd.

_state_icon = {              # HJLK "icon" is sometimes the name of the PICTURE in statuslcd.py !!
      'pause' : '\x00 '
    , 'play'  : '\x01 '
    , 'stop'  : '\x02 '
}_

I have extensively tested this with my 20x4 lcd, but not with a 16x2. 

@rern: what should I do next, to make these changes "official"?

**Reply by rern** [reply]
@HJLK,

Thanks a lot for looking into this.
I've revised the whole thing and will be released with next update soon.

In the mean time, as I believe you can manage if things go wrong, please try the UPDATE branch:
- Addons > rAudio > Long-press `Update` button > OK to the `UPDATE` branch

**Comment by HJLK** [comment]
Hello @Rern, and a good day to you,

I ditd the UPDATE, and rebooted. My LCD was dead again. I'm willing to 
dive into this, but only with your consent. As I can see that you 
modified a lot of files I will wait with "diving" until you release the 
UPDATE.

Is this OK with you?

Greetings from HJLK,

**Reply by rern** [reply]
@HJLK,

The UPDATE branch has been revised to fix it. Please try the Long-press > UPDATE again.

**Comment by HJLK** [comment]
Hellp @Rern, I just did. Made no difference, even after reboot. I checked the **initial** values in /srv/http/data/system/lcdcharconf.py:

_cols=20 charmap=A02 inf=i2c address=39 chip=PCF8574 backlight=False_

This will not work in python import... After choosing System, and modifying the CharLCD params, behold: 

_cols=20
charmap="20"
inf="A02"
address=39
chip="PCF8574"
backlight=False_

There is something wrong here: **charmap** should be "A02", and **inf** should be "i2c". When I corrected this manually and ran lcdchar.py it worked... even with a running time (for radio). It seems that did the trick! I see running time in album tracks, and radio. NICE!
I will continue testing, but so far CharLCD 20x4 looks GOOD!

**Comment by HJLK** [comment]
Testing the UPDATE-branch:

1) CharLCD 20x4:
- initialize/update settings via System menu has problems (see above)
- with album tracks the state 'pause' does not show on lcd. Lcd is refreshed, but shows 'play', and time keeps running.
- with power off: logo stays on lcd. This is ok, but me thinks backlight should be turned off... 
- when you power-off the player while "playing" the lcd flashes the logo, and then returns to the last song played, with the stop-icon.

2) volume control
- control in app && browsergui does not work well: begins 100 and does not follow volume via rotary encoder.
- Mute does not work

3) Boot:
- shutdown.sh is not activated. systemctl status startup shows:


 startup.service - rAudio Startup
     Loaded: loaded (/etc/systemd/system/startup.service; enabled; preset: disabled)
     Active: failed (Result: exit-code) since Sat 2023-04-22 09:49:15 CEST; 30min ago
    Process: 504 ExecStart=/srv/http/bash/startup.sh (code=exited, status=2)
   Main PID: 504 (code=exited, status=2)
        CPU: 3.863s

Apr 21 17:53:00 rAudio systemd[1]: Starting rAudio Startup...
Apr 21 17:53:00 rAudio startup.sh[524]: bsdtar: Error opening archive: Failed to open 'backup.gz'
Apr 21 17:53:00 rAudio startup.sh[527]: /srv/http/bash/startup.sh: line 30: <code>/boot/backup.gz</code> is not rAudio backup.: No such file or directory
Apr 22 09:49:07 rAudio startup.sh[657]: /srv/http/bash/cmd.sh: line 440: /srv/http/data/system/equalizer.json: No such file or directory
Apr 22 09:49:15 rAudio startup.sh[504]: /srv/http/bash/startup.sh: line 130: syntax error near unexpected token `then'
Apr 22 09:49:15 rAudio startup.sh[504]: /srv/http/bash/startup.sh: line 130: `\if [[ $connected ]]; then'
Apr 22 09:49:15 rAudio systemd[1]: startup.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Apr 22 09:49:15 rAudio systemd[1]: startup.service: Failed with result 'exit-code'.
Apr 22 09:49:15 rAudio systemd[1]: Failed to start rAudio Startup.
Apr 22 09:49:15 rAudio systemd[1]: startup.service: Consumed 3.863s CPU time.
~

- Features - Autoplay does not work.

So far, so good!

**Comment by HJLK** [comment]
@rern: to display the 'pause' situation on LCD, you can modify lcdchar.py at line 201:

_# if not elapsed: sys.exit()
if state != 'play': # HJLK: only when playing we do refresh last line per second
	lcd.close()
	sys.exit()_

The loop will now only run when state = 'play'

**Comment by HJLK** [comment]
@rern: in lcdchar.py a Python global variable _backlight_ is defined in _def backlightOff()_, It is called in line 177 for a 2-line LCD, and state != 'play', and in line 199 for all LCD-types.
Has this to do with the settings for the LCD, in combination with the sleep(60), imported from lcdcharconf.py? 

Is it your intention to dim the backlight of the LCD in non-play mode, but only after _waiting_ for 60 seconds?

**Comment by HJLK** [comment]
@rern: I updated to the latest UPDATE branch @17:00 26-4. LCD gets clobbered because of not killing "all" lcdchar.py in status-push.sh. It is modified to "killProcess lcdchar", which does **not** do the job. 
I think I'll wait to do further testing until you are done with the coming release. When do you expect this?

**Comment by HJLK** [comment]
@rern: I saw that you modified status-push.sh back. Yet sometimes the lcd is clobbering again, which means there are 2 occurences of lcdchar.py active. I saw that killall lcdchar.py (manually!) took about 1 second! So it is indeed possible that the "new" lcdchar.py is active before the old one is terminated. I modified this to: "killall -w lcdchar.py &> /dev/null" which will wait for the process to die. Maybe you can use this?

**Comment by HJLK** [comment]
@rern: Greetings to you. I noticed that you made some fundamental changes (using systemd etc) for stopping/starting scripts in the UPDATE-branch.
Can you please let me know when UPDATE-branch is ready for testing? 
By the way: I noticed that "Addons - Update"-"long press" does not work anymore for me.
